src/back/hlsl/writer.rs

```diff
@@ -2688,13 +2722,231 @@ impl<'a, W: fmt::Write> super::Writer<'a, W> {
+                                write!(self.out, "((")?;
+                                self.write_expr(module, arg, func_ctx)?;
+                                write!(self.out, " >> ")?;
+                                self.write_expr(module, offset, func_ctx)?;
+                                write!(self.out, ") & (")?;
+                                self.write_expr(module, count, func_ctx)?;
+                                write!(self.out, " == {scalar_width}u ? {scalar_max}u : ((1u << ")?;
+                                self.write_expr(module, count, func_ctx)?;
+                                write!(self.out, ") - 1)))")?;
+                            } else {
+                                // Signed
+                                // (count == 0 ? 0 : (e << (32 - count - offset)) >> (32 - count))
+                                write!(self.out, "(")?;
+                                self.write_expr(module, count, func_ctx)?;
+                                write!(self.out, " == 0 ? 0 : (")?;
+                                self.write_expr(module, arg, func_ctx)?;
+                                write!(self.out, " << ({scalar_width} - ")?;
+                                self.write_expr(module, count, func_ctx)?;
+                                write!(self.out, " - ")?;
+                                self.write_expr(module, offset, func_ctx)?;
+                                write!(self.out, ")) >> ({scalar_width} - ")?;
+                                self.write_expr(module, count, func_ctx)?;
+                                write!(self.out, "))")?;
+                            }
+                        }
+                    }
+                    Function::InsertBits => {
+                        // e: T,
+                        // newbits: T,
+                        // offset: u32,
+                        // count: u32
+                        // returns T
+                        // T is i32, u32, vecN<i32>, or vecN<u32>
+                        if let (Some(newbits), Some(offset), Some(count)) = (arg1, arg2, arg3) {
+                            let inner = func_ctx.info[expr].ty.inner_with(&module.types);
+                            let scalar_width = inner.scalar_width().unwrap_or(32);
+                            let scalar_max: u64 = match scalar_width {
+                                8 => 0xff,
+                                16 => 0xffff,
+                                32 => 0xffffffff,
+                                64 => 0xffffffffffffffff,
+                                _ => {
+                                    return Err(Error::Unimplemented(format!(
+                                        "write_expr_math insert_bits for scalar_width {}",
+                                        scalar_width
+                                    )))
+                                }
+                            };
```

Comment:
I think we can just assume 32bit here, since that's all we support right now (same for `ExtractBits`).
Thanks for considering this though!

Reply:
Should I leave `scalar_max` as a variable, or should I replace all the places where I use `scalar_max` with 0xFFFFFFFF?

Reply:
Let's keep them as variables.